File: 04/main.js
const fs = require('node:fs');
const crypto = require("crypto");

function find_number(key, zeroes) {
    const search_target = "0".repeat(zeroes);
    let number = 0;
    while (true) {
        let result = crypto.createHash("md5").update(`${key}${number}`).digest("hex")
        if (result.startsWith(search_target)) return number;
        number++;
    }
}

fs.readFile('input.txt', 'utf8', (err, data) => {
    const key = data.trim()
    console.log(`Part1: ${find_number(key, 5)}`);
    console.log(`Part2: ${find_number(key, 6)}`);
});

File: 04/main.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func find_number(key string, zeroes int) int {
	search_target := strings.Repeat("0", zeroes)
	number := 0
	for {
		hash := md5.Sum([]byte(key + strconv.Itoa(number)))
		hex := hex.EncodeToString(hash[:])
		if strings.HasPrefix(hex, search_target) {
			return number
		}
		number++
	}
}

func main() {
	b, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print(err)
		return
	}
	key := string(b)
	fmt.Printf("Part1: %d\n", find_number(key, 5))
	fmt.Printf("Part2: %d\n", find_number(key, 6))
}
